components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java

```diff
@@ -1944,39 +1944,63 @@ private void recoverFromConAppKeyConstraintViolation(String accessToken, String
             AccessTokenDO latestActiveToken = getLatestAccessTokenByState(connection, consumerKey,
                     accessTokenDO.getAuthzUser(), userStoreDomain,
                     OAuth2Util.buildScopeString(accessTokenDO.getScope()), true);
+            OauthTokenIssuer oauthTokenIssuer = OAuth2Util.getOAuthTokenIssuerForOAuthApp(consumerKey);
+
+            // For JWT tokens, always issue a new token expiring the existing token.
+            if (oauthTokenIssuer.renewAccessTokenPerRequest()) {
+
+                updateAccessTokenState(connection, latestActiveToken.getTokenId(), OAuthConstants.TokenStates
+                        .TOKEN_STATE_EXPIRED, UUID.randomUUID().toString(), userStoreDomain);
+                // Update token issued time make this token as latest token & try to store it again.
+                accessTokenDO.setIssuedTime(new Timestamp(new Date().getTime()));
+                insertAccessToken(accessTokenDO.getAccessToken(), consumerKey, accessTokenDO, connection,
+                        userStoreDomain, retryAttemptCounter);
+            } else {
+                if (latestActiveToken != null) {
+                    if (OAuth2Util.getAccessTokenExpireMillis(latestActiveToken) > 0 && (latestNonActiveToken == null ||
+                            latestActiveToken.getIssuedTime().after(latestNonActiveToken.getIssuedTime()))) {
+                        // In here we can use existing token since we have a synchronised communication
+                        accessTokenDO.setTokenId(latestActiveToken.getTokenId());
+                        accessTokenDO.setAccessToken(latestActiveToken.getAccessToken());
+                        accessTokenDO.setRefreshToken(latestActiveToken.getRefreshToken());
+                        accessTokenDO.setIssuedTime(latestActiveToken.getIssuedTime());
+                        accessTokenDO.setRefreshTokenIssuedTime(latestActiveToken.getRefreshTokenIssuedTime());
+                        accessTokenDO.setValidityPeriodInMillis(latestActiveToken.getValidityPeriodInMillis());
+                        accessTokenDO.setRefreshTokenValidityPeriodInMillis(
+                                latestActiveToken.getRefreshTokenValidityPeriodInMillis());
+                        accessTokenDO.setTokenType(latestActiveToken.getTokenType());
+                        log.info("Successfully recovered 'CON_APP_KEY' constraint violation with the attempt : "
+                                + retryAttemptCounter);
+
+                    } else if (!(OAuth2Util.getAccessTokenExpireMillis(latestActiveToken) > 0)) {
```

Comment:
Can't we use this as OAuth2Util.getAccessTokenExpireMillis(latestActiveToken) <= 0)

Reply:
We should only check whether it is equal to 0 because for the infinite tokens the time to expire will be -1